writeengine/wrapper/writeengine.cpp

```diff
@@ -220,6 +218,178 @@ void WriteEngineWrapper::findSmallestColumn(uint32_t& colId, ColStructList colSt
     }
 }
 
+/** @brief Fetch values from arrays into VT-types references casting arrays to (element type) ET-typed arrays. */
+template <typename VT, typename ET>
+void fetchNewOldValues(VT& value, VT& oldValue, const void* array, const void* oldArray, size_t i, size_t totalNewRow)
+{
+    const ET* eArray = static_cast<const ET*>(array);
+    const ET* oldEArray = static_cast<const ET*>(oldArray);
+    if (eArray)
+    {
+        value = eArray[i < totalNewRow ? i : 0];
+    }
+    if (oldEArray)
+    {
+        oldValue = oldEArray[i];
+    }
+}
+
+/**
+ * There can be case with missing valArray (delete), missing oldValArray (insert) and when
+ * both arrays are present (update).
+ */
+void WriteEngineWrapper::updateMaxMinRange(const size_t totalNewRow, const size_t totalOldRow,
+                                const execplan::CalpontSystemCatalog::ColType& cscColType,
+                                const ColType colType,
+                                const void* valArrayVoid, const void* oldValArrayVoid,
+                                BRM::CPInfo* maxMin, bool canStartWithInvalidRange)
+{
+    if (!maxMin)
+    {
+        return ;
+    }
+    bool isUnsigned = false; // TODO: should change with type.
+    switch (colType)
+    {
+        case WR_UBYTE:
+        case WR_USHORT:
+        case WR_UINT:
+        case WR_ULONGLONG:
+        {
+            isUnsigned = true;
+            break;
+        }
+        default:
+            break;
+    }
+    if (!canStartWithInvalidRange)
+    {
+        // check if range is invalid, we can't update it.
+        if (isUnsigned)
+	{
+            if (static_cast<uint64_t>(maxMin->max) < static_cast<uint64_t>(maxMin->min))
+            {
+                return;
+            }
+	}
+	else
+	{
+            if (maxMin->max < maxMin->min)
+            {
+                return;
+            }
+	}
+    }
+    size_t i;
+    for (i = 0; i < totalOldRow; i++) {
+	int64_t value, oldValue;
+        uint64_t uvalue, oldUValue;
+
+	// fetching. May not assign value or oldValue variables when array is not present.
+        switch (colType)
```

Comment:
Despite the fact that CPU branch prediction should eliminate the negative effect of a switch called in a loop is it possible to move the switch above the loop?

Reply:
I also don't see int128_t support here.

Reply:
Now we do support wide decimals.

Reply:
Plz put the switch statement outside of the loop using std::function wrapper to save needed fetchNewOldValues<T, T> and call it later in the loop.